Client/Frontend/UserContent/UserScripts/AllFrames/AtDocumentStart/RewardsReporting.js

```diff
@@ -0,0 +1,100 @@
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+"use strict";
+
+if (webkit.messageHandlers.rewardsReporting) {
+    install();
+}
+
+function install() {
+    function sendMessage(method, url, data, referrerUrl) {
+        webkit.messageHandlers.rewardsReporting.postMessage({
+            method: method === undefined ? "GET" : method,
+            url: url,
+            data: data === undefined ? null : data,
+            referrerUrl: referrerUrl === undefined ? null : referrerUrl,
+        });
+    }
+    
+    let originalOpen = XMLHttpRequest.prototype.open;
+    let originalSend = XMLHttpRequest.prototype.send;
+    let originalFetch = window.fetch;
+    let originalSendBeacon = navigator.sendBeacon;
+    let originalImageSrc = Object.getOwnPropertyDescriptor(Image.prototype, "src");
+    let mutationObserver = null;
+    
+    XMLHttpRequest.prototype.open = function(method, url) {
+        const listener = function() {
+            sendMessage(this._method, this.responseURL === null ? this._url : this.responseURL, this._data, this._ref);
+        };
+        this._method = method;
+        this._url = url;
+        this.addEventListener('load', listener, true);
+        this.addEventListener('error', listener, true);
+        return originalOpen.apply(this, arguments);
+    };
+    
+    XMLHttpRequest.prototype.send = function(body) {
+        this._ref = null;
+        this._data = body;
+        if (body instanceof Document) {
+            this._ref = body.referrer;
+            this._data = null;
+        }
+        return originalSend.apply(this, arguments);
+    };
+
+    window.fetch = function(resource, options) {
+        const args = arguments
+        const url = resource instanceof Request ? resource.url : resource
+
+        return new Promise(function(resolve, reject) {
+            originalFetch.apply(this, args)
+            .then(function(response) {
+                sendMessage(options.method, url, options.body, options.referrer);
+                resolve(response);
+            })
+            .catch(function(error) {
+                sendMessage(options.method, url, options.body, options.referrer);
+                reject(error);
+            })
+        });
+    };
+
+    navigator.sendBeacon = function(url, data) {
+        sendMessage("POST", url, data);
+        return originalSendBeacon.apply(this, arguments);
+    };
+    
+    delete Image.prototype.src;
+    Object.defineProperty(Image.prototype, "src", {
+      get: function() {
+        return originalImageSrc.get.call(this);
+      },
+      set: function(value) {
+        const listener = function() {
+            sendMessage("GET", this.src);
+        };
+        this.addEventListener('load', listener, true);
+        this.addEventListener('error', listener, true);
+        originalImageSrc.set.call(this, value);
+      }
+    });
+
+    mutationObserver = new MutationObserver(function(mutations) {
```

Comment:
FWIW i think we got rid of mutation observers in browser-laptop a while ago due to performance issues

Reply:
I'm not sure if there'll be a better way for us to do this because desktop reports all script loads: https://github.com/brave/brave-core/blob/5a217978eee7639e7d8b9b755a4af11504698dda/browser/brave_rewards/rewards_tab_helper.cc#L81

Reply:
@NejcZdovc can you confirm why we need to check every script/XHR/etc. load in brave-core rewards to see if it is a media type?

i glanced through the code and it seems this is only necessary to check if the load is a twitch/youtube/vimeo request? 

i would rather we not override the prototypes and add mutation observers on every page just to see if they are twitch/youtube/vimeo loads

Reply:
we need to send every url into native ledger, which has business logic to check if twitch/youtube/etc. If we don't detect any media publisher we will ignore request

Reply:
@NejcZdovc would it be acceptable to only send these requests if the top-level (or iframe-level) page is a twitch/youtube/etc. page? that would limit the potential impact of this change.

Reply:
problem is that then business logic is moved from native ledger (shared lib) to platform specific code. This means that if we add new publisher every platform need to add support for it instead of just pulling the latest shared lib

Reply:
@NejcZdovc a somewhat separate question is - why does native ledger need to inspect all these subresource loads (XHRs, images) instead of just checking the top-level or iframe-level URL to see if it is twitch/youtube/etc?

Reply:
I already left comment in the code that I think we only need XHR and nothing else

Reply:
I think you'll need at least Image as well for YouTube since that's how the load is classified for `watchtime`. Fetch and beacon API calls probably run through XHR in desktop for twitch and vimeo. As far as I've seen none of the publishers currently require script monitoring.

Reply:
but AFAICT, we only need to report those XHRs on youtube.com and youtube embeds. so it seems preferable for me to only run this content script if the frame URL matches youtube/vimeo/twitch/etc. instead of on all pages.

i agree this is not ideal from a code-reuse perspective, but i think it's probably worth it to limit the potential webcompat (and maybe security) impacts of this overriding